File: frontend/src/types/temporal.ts
// T2 시계열 인사이트 타입 정의

export type CompareMode = 'products' | 'periods' | 'categories';

export interface TemporalPoint {
  date: string;        // YYYY-MM-DD
  count: number;       // VOC 건수
  sent_avg: number;    // 평균 감성 점수 (-1 ~ 1)
}

export interface TemporalSeries {
  key: string;         // product_code | period_label | category_code
  label: string;       // 사용자 노출용
  data: TemporalPoint[];
}

export interface TimelineEvent {
  date: string;        // YYYY-MM-DD
  title: string;       // 이벤트 타이틀
  category?: string;   // launch | update | issue | etc.
  product_code?: string;
}

export interface Changepoint {
  date: string;
  series_key: string;
  delta: number;       // 변화 폭
  reason?: string;     // 추정 사유
}

export interface TemporalSeriesResponse {
  mode: CompareMode;
  series: TemporalSeries[];
  events: TimelineEvent[];
  changepoints: Changepoint[];
}

// LLM Narrative 요청/응답
export interface LLMNarrativeCitation {
  source_url: string;
  title?: string;
  snippet?: string;
}

export interface LLMNarrativeRequest {
  mode: CompareMode;
  series_keys: string[];
  date_start: string;
  date_end: string;
  // 차트에서 본 데이터를 LLM이 분석할 수 있도록 컨텍스트로 첨부
  context?: {
    series: TemporalSeries[];
    events: TimelineEvent[];
    changepoints: Changepoint[];
  };
}

export interface LLMNarrativeResponse {
  narrative: string;            // Markdown 본문 (한국어)
  citations: LLMNarrativeCitation[];
  model?: string;               // 응답 모델명 (qwen2.5:7b 등)
  generated_at?: string;        // ISO timestamp
}

File: backend/app/schemas/temporal.py
"""
T2 시계열+LLM 분석 Pydantic schemas (P2-3).

엔드포인트:
- GET  /api/v1/analytics/temporal-series
- GET  /api/v1/analytics/temporal-compare
- POST /api/v1/analytics/llm-narrative
"""
from __future__ import annotations

from typing import List, Optional, Literal, Dict, Any

from pydantic import BaseModel, Field


# ── 공통 ─────────────────────────────────────────────────────────────
class SeriesPoint(BaseModel):
    date: str                       # 'YYYY-MM-DD'
    count: int
    sent_avg: float                 # -1 ~ 1
    neg_rate: float                 # 0 ~ 100
    pos_rate: float                 # 0 ~ 100


class TimelineEvent(BaseModel):
    date: str                       # 'YYYY-MM-DD'
    type: str                       # 'release' | 'update' | 'incident' | 'pr'
    title: str
    product_code: Optional[str] = None
    source_url: Optional[str] = None


class ChangePoint(BaseModel):
    date: str
    metric: Literal["count", "sent_avg"]
    magnitude: float                # |delta| (절대값)
    direction: Literal["up", "down"]


# ── 1) temporal-series ──────────────────────────────────────────────
class TemporalSeriesResponse(BaseModel):
    series: List[SeriesPoint]
    events: List[TimelineEvent] = Field(default_factory=list)
    changepoints: List[ChangePoint] = Field(default_factory=list)
    meta: Dict[str, Any] = Field(default_factory=dict)


# ── 2) temporal-compare ─────────────────────────────────────────────
class CompareSeries(BaseModel):
    key: str
    points: List[SeriesPoint]


class DiffPoint(BaseModel):
    date: str
    delta_count: int                # a - b
    delta_sent: float               # a - b


class TemporalCompareResponse(BaseModel):
    mode: Literal["products", "periods", "categories"]
    a: CompareSeries
    b: CompareSeries
    diff: List[DiffPoint]


# ── 3) llm-narrative ────────────────────────────────────────────────
class LLMNarrativeRequest(BaseModel):
    series_payload: Dict[str, Any]
    lang: str = "ko"


class NarrativeCitation(BaseModel):
    event_date: str
    source_url: Optional[str] = None
    title: Optional[str] = None


class LLMNarrativeResponse(BaseModel):
    summary: str
    citations: List[NarrativeCitation] = Field(default_factory=list)
    cached: bool = False
    provider: Optional[str] = None


__all__ = [
    "SeriesPoint",
    "TimelineEvent",
    "ChangePoint",
    "TemporalSeriesResponse",
    "CompareSeries",
    "DiffPoint",
    "TemporalCompareResponse",
    "LLMNarrativeRequest",
    "NarrativeCitation",
    "LLMNarrativeResponse",
]
